File: network/src/main/java/info/metadude/android/eventfahrplan/network/fetching/FetchFahrplan.java
package info.metadude.android.eventfahrplan.network.fetching;

import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.net.UnknownServiceException;

import javax.net.ssl.SSLException;
import javax.net.ssl.SSLHandshakeException;

import info.metadude.android.eventfahrplan.commons.logging.Logging;
import okhttp3.Call;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class FetchFahrplan {

    public interface OnDownloadCompleteListener {

        void onGotResponse(@NonNull FetchScheduleResult fetchScheduleResult);
    }

    @NonNull
    private final Logging logging;

    @Nullable
    private FetchFahrplanTask task;

    private OnDownloadCompleteListener listener;

    public FetchFahrplan(@NonNull Logging logging) {
        this.logging = logging;
        task = null;
    }

    public void fetch(@NonNull OkHttpClient okHttpClient, String url, String eTag) {
        task = new FetchFahrplanTask(okHttpClient, logging, this.listener);
        task.execute(url, eTag);
    }

    public void cancel() {
        if (task != null) {
            task.cancel(true);
        }
    }

    public void setListener(OnDownloadCompleteListener listener) {
        this.listener = listener;
        if (task != null) {
            task.setListener(listener);
        }
    }
}

class FetchFahrplanTask extends AsyncTask<String, Void, HttpStatus> {

    private static final String EMPTY_RESPONSE_STRING = "";

    private final OkHttpClient okHttpClient;

    @NonNull
    private final Logging logging;

    private String responseStr;

    @NonNull
    private String eTagStr = "";

    private final String LOG_TAG = "FetchFahrplan";

    private FetchFahrplan.OnDownloadCompleteListener listener;

    private boolean completed;

    private HttpStatus status;

    @NonNull
    private String host = "";

    @NonNull
    private String exceptionMessage = "";

    FetchFahrplanTask(
            @NonNull OkHttpClient okHttpClient,
            @NonNull Logging logging,
            FetchFahrplan.OnDownloadCompleteListener listener
    ) {
        this.okHttpClient = okHttpClient;
        this.logging = logging;
        this.listener = listener;
        this.completed = false;
    }

    public void setListener(FetchFahrplan.OnDownloadCompleteListener listener) {
        this.listener = listener;

        if (completed && listener != null) {
            notifyActivity();
        }
    }

    @Override
    protected HttpStatus doInBackground(String... args) {
        String url = args[0];
        String eTag = args[1];

        //noinspection ConstantConditions
        host = Uri.parse(url).getHost();
        if (host == null) {
            throw new NullPointerException("Host is null for url = '" + url + "'");
        }

        return fetch(url, eTag);
    }

    protected void onCancelled() {
        logging.d(LOG_TAG, "Fetch cancelled");
    }

    protected void onPostExecute(HttpStatus status) {
        completed = true;
        this.status = status;

        if (listener != null) {
            notifyActivity();
        }
    }

    private void notifyActivity() {
        if (status == HttpStatus.HTTP_OK) {
            logging.d(LOG_TAG, "Fetch done successfully");
            listener.onGotResponse(new FetchScheduleResult(status, responseStr, eTagStr, host, exceptionMessage));
        } else {
            logging.d(LOG_TAG, "Fetch failed");
            listener.onGotResponse(new FetchScheduleResult(status, EMPTY_RESPONSE_STRING, eTagStr, host, exceptionMessage));
        }
        completed = false; // notify only once
    }

    private HttpStatus fetch(String url, String eTag) {
        logging.d(LOG_TAG, url);
        logging.d(LOG_TAG, "ETag: '" + eTag + "'");
        Request.Builder requestBuilder = new Request.Builder()
                .url(url);

        if (!TextUtils.isEmpty(eTag)) {
            requestBuilder.addHeader("If-None-Match", eTag);
        }

        Response response;
        try {
            Call call = okHttpClient.newCall(requestBuilder.build());
            response = call.execute();
        } catch (SSLException e) {
            setExceptionMessage(e);
            customizeExceptionMessage(e);
            e.printStackTrace();
            return HttpStatus.HTTP_LOGIN_FAIL_UNTRUSTED_CERTIFICATE;
        } catch (SocketTimeoutException e) {
            return HttpStatus.HTTP_CONNECT_TIMEOUT;
        } catch (UnknownHostException e) {
            e.printStackTrace();
            return HttpStatus.HTTP_DNS_FAILURE;
        } catch (UnknownServiceException e) {
            e.printStackTrace();
            return HttpStatus.HTTP_CLEARTEXT_NOT_PERMITTED;
        } catch (IOException e) {
            e.printStackTrace();
            return HttpStatus.HTTP_COULD_NOT_CONNECT;
        }

        int statusCode = response.code();
        if (statusCode == 304) {
            return HttpStatus.HTTP_NOT_MODIFIED;
        }

        if (statusCode != 200) {
            logging.e(LOG_TAG, "Error " + statusCode + " while retrieving XML data");
            if (statusCode == 401) {
                return HttpStatus.HTTP_WRONG_HTTP_CREDENTIALS;
            }
            if (statusCode == 404) {
                return HttpStatus.HTTP_NOT_FOUND;
            }
            return HttpStatus.HTTP_COULD_NOT_CONNECT;
        }

        //noinspection ConstantConditions
        eTagStr = response.header("ETag");
        eTagStr = eTagStr == null ? "" : eTagStr;
        if (!eTagStr.isEmpty()) {
            logging.d(LOG_TAG, "ETag: '" + eTagStr + "'");
        } else {
            logging.d(LOG_TAG, "ETag missing?");
        }

        try {
            //noinspection ConstantConditions
            responseStr = response.body().string();
        } catch (NullPointerException | IOException e) {
            return HttpStatus.HTTP_CANNOT_PARSE_CONTENT;
        } finally {
            if (response.body() != null) {
                response.body().close();
            }
        }

        return HttpStatus.HTTP_OK;
    }

    @SuppressWarnings("ConstantConditions")
    private void setExceptionMessage(SSLException exception) {
        if (exception.getCause() == null) {
            exceptionMessage = exception.getMessage();
        } else {
            if (exception.getCause().getCause() == null) {
                exceptionMessage = exception.getCause().getMessage();
            } else {
                exceptionMessage = exception.getCause().getCause().getMessage();
            }
        }
        if (exceptionMessage == null) {
            exceptionMessage = "";
        }
    }

    private void customizeExceptionMessage(@NonNull SSLException exception) {
        if (exception instanceof SSLHandshakeException && Build.VERSION.SDK_INT == Build.VERSION_CODES.N) {
            // See https://github.com/EventFahrplan/EventFahrplan/issues/431
            exceptionMessage += "\n\nPlease note that server certificates using elliptic curves " +
                    "with a length > 256 bits are not supported on Android 7.0. This might cause " +
                    "this error.";
        }
    }

}

File: network/src/main/java/info/metadude/android/eventfahrplan/network/fetching/FetchFahrplan.kt
package info.metadude.android.eventfahrplan.network.fetching

import android.net.Uri
import android.os.AsyncTask
import android.os.Build

import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.net.UnknownServiceException

import javax.net.ssl.SSLException
import javax.net.ssl.SSLHandshakeException

import info.metadude.android.eventfahrplan.commons.logging.Logging
import info.metadude.android.eventfahrplan.network.models.HttpHeader

import okhttp3.OkHttpClient
import okhttp3.Request

internal class FetchFahrplan(private val logging: Logging) {

    private lateinit var task: FetchFahrplanTask
    private lateinit var onFetchScheduleResult: (fetchScheduleResult: FetchScheduleResult) -> Unit

    fun fetch(okHttpClient: OkHttpClient, url: String, httpHeader: HttpHeader) {
        task = FetchFahrplanTask(okHttpClient, logging, onFetchScheduleResult)
        task.execute(url, httpHeader.eTag, httpHeader.lastModified)
    }

    fun cancel() {
        if (::task.isInitialized) {
            task.cancel(true)
        }
    }

    fun setListener(listener: (fetchScheduleResult: FetchScheduleResult) -> Unit) {
        onFetchScheduleResult = listener
        if (::task.isInitialized) {
            task.setListener(listener)
        }
    }
}

internal class FetchFahrplanTask(

    private val okHttpClient: OkHttpClient,
    private val logging: Logging,
    private var onFetchScheduleResult: (fetchScheduleResult: FetchScheduleResult) -> Unit

) : AsyncTask<String, Void, HttpStatus>() {

    private companion object {
        const val EMPTY_RESPONSE_STRING = ""
        const val LOG_TAG = "FetchFahrplan"
        const val HTTP_HEADER_NAME_ETAG = "ETag"
        const val HTTP_HEADER_NAME_IF_NONE_MATCH = "If-None-Match"
        const val HTTP_HEADER_NAME_LAST_MODIFIED = "Last-Modified"
        const val HTTP_HEADER_NAME_IF_MODIFIED_SINCE = "If-Modified-Since"
    }

    private var responseStr = EMPTY_RESPONSE_STRING
    private var eTagStr = ""
    private var lastModifiedStr = ""
    private var completed = false
    private lateinit var status: HttpStatus
    private var host = ""
    private var exceptionMessage = ""

    fun setListener(listener: (fetchScheduleResult: FetchScheduleResult) -> Unit) {
        onFetchScheduleResult = listener
        if (completed) {
            notifyActivity()
        }
    }

    @Deprecated("Deprecated in Java")
    override fun doInBackground(vararg args: String?): HttpStatus {
        val url = args[0]!!
        val eTag = args[1]!!
        val lastModified = args[2]!!
        host = Uri.parse(url).host ?: throw NullPointerException("Host is null for url = '$url'")
        return fetch(url, eTag, lastModified)
    }

    @Deprecated("Deprecated in Java")
    override fun onCancelled() {
        logging.d(LOG_TAG, "Fetch cancelled")
    }

    @Deprecated("Deprecated in Java")
    override fun onPostExecute(status: HttpStatus) {
        completed = true
        this.status = status
        notifyActivity()
    }

    private fun notifyActivity() {
        if (status == HttpStatus.HTTP_OK) {
            logging.d(LOG_TAG, "Fetch done successfully")
            onFetchScheduleResult(
                FetchScheduleResult(status, responseStr, HttpHeader(eTagStr, lastModifiedStr), host, exceptionMessage)
            )
        } else {
            logging.d(LOG_TAG, "Fetch failed")
            onFetchScheduleResult(
                FetchScheduleResult(
                    status, EMPTY_RESPONSE_STRING, HttpHeader(eTagStr, lastModifiedStr), host, exceptionMessage
                )
            )
        }
        completed = false // notify only once
    }

    private fun fetch(url: String, eTag: String, lastModified: String): HttpStatus {
        logging.d(LOG_TAG, url)
        logging.d(LOG_TAG, "$HTTP_HEADER_NAME_ETAG: '$eTag'")
        logging.d(LOG_TAG, "$HTTP_HEADER_NAME_LAST_MODIFIED: '$lastModified'")
        val requestBuilder = Request.Builder().apply {
            url(url)
            if (eTag.isNotEmpty()) {
                addHeader(HTTP_HEADER_NAME_IF_NONE_MATCH, eTag)
            }
            if (lastModified.isNotEmpty()) {
                addHeader(HTTP_HEADER_NAME_IF_MODIFIED_SINCE, lastModified)
            }
        }

        val response = try {
            okHttpClient.newCall(requestBuilder.build()).execute()
        } catch (e: SSLException) {
            setExceptionMessage(e)
            customizeExceptionMessage(e)
            e.printStackTrace()
            return HttpStatus.HTTP_LOGIN_FAIL_UNTRUSTED_CERTIFICATE
        } catch (e: SocketTimeoutException) {
            return HttpStatus.HTTP_CONNECT_TIMEOUT
        } catch (e: UnknownHostException) {
            e.printStackTrace()
            return HttpStatus.HTTP_DNS_FAILURE
        } catch (e: UnknownServiceException) {
            e.printStackTrace()
            return HttpStatus.HTTP_CLEARTEXT_NOT_PERMITTED
        } catch (e: IOException) {
            e.printStackTrace()
            return HttpStatus.HTTP_COULD_NOT_CONNECT
        }
        val statusCode = response.code

        if (statusCode == 304) {
            return HttpStatus.HTTP_NOT_MODIFIED
        }

        if (statusCode != 200) {
            logging.e(LOG_TAG, "Error $statusCode while retrieving XML data")
            if (statusCode == 401) {
                return HttpStatus.HTTP_WRONG_HTTP_CREDENTIALS
            }
            return if (statusCode == 404) {
                HttpStatus.HTTP_NOT_FOUND
            } else {
                HttpStatus.HTTP_COULD_NOT_CONNECT
            }
        }

        eTagStr = response.header(HTTP_HEADER_NAME_ETAG).orEmpty()
        if (eTagStr.isEmpty()) {
            logging.d(LOG_TAG, "$HTTP_HEADER_NAME_ETAG is missing.")
        } else {
            logging.d(LOG_TAG, "$HTTP_HEADER_NAME_ETAG: '$eTagStr'")
        }

        lastModifiedStr = response.header(HTTP_HEADER_NAME_LAST_MODIFIED).orEmpty()
        if (lastModifiedStr.isEmpty()) {
            logging.d(LOG_TAG, "$HTTP_HEADER_NAME_LAST_MODIFIED is missing")
        } else {
            logging.d(LOG_TAG, "$HTTP_HEADER_NAME_LAST_MODIFIED: '$lastModifiedStr'")
        }

        responseStr = try {
            response.body!!.string()
        } catch (e: NullPointerException) {
            return HttpStatus.HTTP_CANNOT_PARSE_CONTENT
        } catch (e: IOException) {
            return HttpStatus.HTTP_CANNOT_PARSE_CONTENT
        } finally {
            response.body?.close()
        }
        return HttpStatus.HTTP_OK
    }

    private fun setExceptionMessage(exception: SSLException) {
        fun getExceptionMessage(cause: Throwable?, message: String?): String? =
            if (cause == null) message else getExceptionMessage(cause.cause, cause.message)

        exceptionMessage = getExceptionMessage(exception.cause, exception.message).orEmpty()
    }

    private fun customizeExceptionMessage(exception: SSLException) {
        if (exception is SSLHandshakeException && Build.VERSION.SDK_INT == Build.VERSION_CODES.N) {
            // See https://github.com/EventFahrplan/EventFahrplan/issues/431
            exceptionMessage += "\n\nPlease note that server certificates using elliptic curves " +
                    "with a length > 256 bits are not supported on Android 7.0. This might cause " +
                    "this error."
        }
    }

}
